pintc/src/predicate/analyse/type_check.rs

```diff
@@ -944,139 +939,153 @@ impl Contract {
                 then_expr,
                 else_expr,
                 span,
-            } => self.infer_select_expr(*condition, *then_expr, *else_expr, span),
+            } => Ok(self.infer_select_expr(handler, *condition, *then_expr, *else_expr, span)),
 
-            Expr::Index { expr, index, span } => self.infer_index_expr(*expr, *index, span),
+            Expr::Index { expr, index, span } => {
+                Ok(self.infer_index_expr(handler, *expr, *index, span))
+            }
 
             Expr::TupleFieldAccess { tuple, field, span } => {
-                self.infer_tuple_access_expr(*tuple, field, span)
+                Ok(self.infer_tuple_access_expr(handler, *tuple, field, span))
             }
 
-            Expr::Cast { value, ty, span } => self.infer_cast_expr(*value, ty, span),
+            Expr::Cast { value, ty, span } => Ok(self.infer_cast_expr(handler, *value, ty, span)),
 
             Expr::In {
                 value,
                 collection,
                 span,
-            } => self.infer_in_expr(*value, *collection, span),
+            } => Ok(self.infer_in_expr(handler, *value, *collection, span)),
 
-            Expr::Range { lb, ub, span } => self.infer_range_expr(*lb, *ub, span),
+            Expr::Range { lb, ub, span } => Ok(self.infer_range_expr(handler, *lb, *ub, span)),
 
             Expr::Generator {
                 kind,
                 gen_ranges,
                 conditions,
                 body,
                 span,
-            } => self.infer_generator_expr(kind, gen_ranges, conditions, *body, span),
+            } => Ok(self.infer_generator_expr(handler, kind, gen_ranges, conditions, *body, span)),
```

Comment:
What's a little weird here is that we may be returning `Ok(..)` on something that may fail. Ideally, you want `infer_generator_expr` to return a `Result<_, ErrorEmitted>` and just bubble that up. It's a big change to what you have but I think it's more robust. That way, we can ask the question: "did this method fail?".

Also, for that to work well, you need a scoped handler:
```
handler.scope(|handler| infer_generator_expr(handler, ...))
```
and then, in `infer_generator_expr`, you would add at the end:
```
        if handler.has_errors() {
            Err(handler.cancel());
        } else {
            Ok(())
        }
```
something like that...

Reply:
> That way, we can ask the question: "did this method fail?".

Or we can ignore it if we want.  Flexibility both ways.  Sometimes we can just ask the handler `.has_errors()`? but that isn't necessarily specific enough for all cases.

```
  if handler.has_errors() {
    Err(handler.cancel());
  } else {
    Ok(())
  }
```
We use this often enough that it should probably be a method on `Handler` IMO.

Reply:
After talking with @mohammadfawaz we've determined it isn't possible to return `Result<Inference, ErrorEmitted>` with the current type inference structure. This PR cuts down on unnecessary errors by doing both, returning an `Inference` and emitting an error through `handler.emit_err()`.

We talked about changing the structure to return a `Result<(), ErrorEmitted>` and having `set_ty()` at the end of each inference branch. This would let us both set the type to be the `Inference` and return `ErrorEmitted` if there was a malformed `expr`. However this fails when we have `Inference::Dependant` as we need to bubble the `Inference` up to `type_check()` so that we can check the type of the `dependant expr` first.

In the end we decided to leave the structure as it is and if we feel we need to react to an `err_emit()` in the future, we can address restructuring then.